### ISSUE TITLE
YOLOv8 embeddings

### ISSUE THREAD

**Issue by piotr-rarus** [ultralytics/ultralytics#5361]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello folks,
is it possible to obtain some sort of embeddings from the model during forward pass? My problem is I don't have much compute power to run another img2vec model to get embeddings for every detection denoted by a bounding box. I've been wondering if I could get something similar from yolov8. I can register a hook to get activations but this architecture isn't simple FC at 2nd last layer. I'm not quite sure how yolo does extract predictions from DFL, which is the last layer, nor how to get embeddings for a particular bounding box.

```py
activation = {}
def get_activation(name):
    def hook(model, input, output):
        activation[name] = output.detach().cpu().numpy()
    return hook


model.model._modules["model"]._modules["22"]._modules["dfl"].register_forward_hook(get_activation('photo_b'))
```

**Comment by glenn-jocher** [comment]
@piotr-rarus hello,

Thanks for reaching out. YOLOv8 doesn't inherently provide an option to get embeddings during a forward pass. It's primarily designed for object detection so its main output is prediction tensors that are related to object detection tasks such as bounding boxes, objectness scores, and class probabilities.

However, your idea of using hooks to extract activations from a specific layer is feasible. That said, keep in mind that these activations will not be the same as embeddings you would get from an image-to-vector (img2vec) model. The activations from a convolutional layer in YOLO are feature maps, which are high-dimensional and not readily usable as compact feature vectors the way embeddings are.

Also, the activations you get might not be for a particular bounding box as you mentioned. They are computed for the whole image/frame during the forward pass. Any bounding box-specific information is contained within these full-image feature maps, and extracting meaningful bounding box-specific embeddings might involve more complex manipulations.

As for your code snippet, it seems like you're registering a forward hook for the DFL at the end of the model. While this would give you the final output activations of the model, they are direct detection outputs and may not serve as useful "embeddings" for other tasks.

Remember, if you decide to proceed with this approach to carefully define your interpretation of these "embeddings", and you might need additional customization and methods to suit your application better.

I hope this helps.

Best regards,
Glenn

**Comment by piotr-rarus** [comment]
Thanks for you answer. I'm aware such embeddings wouldn't be of good quality. Can't afford to host yet another model on my modest jetson. I hoped I could get at least some form of representation to help me track objects in subsequent frames.

That snippet was just a hook example to convey the general idea. Have a jolly weekend @glenn-jocher. ;)

**Comment by glenn-jocher** [comment]
Hello @piotr-rarus,

Thank you for your response. I understand your constraints and the need for a more efficient way to track objects across frames.

As you mentioned, using the output activations from certain layers in YOLOv8 has potential for object representation. While these representations might not be as discriminative or compact as embeddings from a dedicated img2vec model, they could still capture useful visual information.

However, note that these outputs are much larger in size and dimension compared to traditional embeddings, which can affect memory usage and computation time. Additionally, these raw features are not necessarily tuned for object tracking tasks and may lack the invariances and discriminative properties that make embeddings effective for such tasks.

Object tracking in subsequent frames also involves considerations beyond simply obtaining embeddings for detected objects. For instance, related tasks could include object matching across frames, dealing with occlusion or missed detections, among others. These complexities would mean that a more comprehensive solution might be required.

Lastly, keep in mind that if you wish to derive embeddings for individual bounding boxes from the feature maps, it may need complex manipulations, as these feature maps are typically computed for entire frames.

Have a great weekend!

Best regards,
Glenn

**Comment by waqas-phc** [comment]
This is good approach,
As we have yolov8 now, does the `modal.embed ` returns embedding of the complete image or just the extracted BB in a torch tensor.

I have to extract the whole image embedding representation which can be used for other Multi Modality down streaming task.
@glenn-jocher

**Comment by waqas-phc** [comment]
Like Getting latent space from SPPF layer and pop the rest of the modal

**Comment by glenn-jocher** [comment]
Hello,

The `model.embed()` function isn't designed to return embeddings for the entire image or specific bounding boxes. YOLOv8 focuses on object detection, and extracting meaningful embeddings for multi-modality tasks would require additional customization. For object position understanding, you might need to implement further processing on the feature maps.

**Comment by waqas-phc** [comment]
Thanks @glenn-jocher 
But what about SPPF layer. Does hook from this layer will be use full in any case.

**Comment by glenn-jocher** [comment]
The SPPF layer can provide feature maps that might be useful for extracting image-level representations, but it requires additional processing to be effective for specific tasks.

**Comment by waqas-phc** [comment]
@glenn-jocher Thanks man ,
I will take a look. Again

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions, feel free to ask.

**Comment by piotr-rarus** [comment]
Kinda surprised to see this thread being resurrected after a year :d
I tried to look for any articles or examples using these features from 2nd last layer for downstream tasks but couldn't find much. @glenn-jocher are you aware of any?

**Comment by glenn-jocher** [comment]
I'm not aware of specific reports using these embeddings for downstream tasks. You might explore academic papers or forums for insights on similar approaches.

**Comment by DeoLeung** [comment]
I was wondering if i can leverage yolo's model to do image similarity search, and from the https://github.com/ultralytics/ultralytics/pull/6320#issuecomment-2406765054 it seems possible to get the embedding. any further document about whether it's suitable for image similarity and if yes, what distance function is recommended?

**Comment by glenn-jocher** [comment]
YOLO models aren't specifically designed for image similarity search, but you can experiment with feature maps from layers like SPPF. For similarity, cosine similarity or Euclidean distance are commonly used. Further documentation on this specific use is limited, so experimentation may be necessary.

**Comment by waqas-phc** [comment]
Hi again @glenn-jocher 
So I am using Yolo11 and taking the image features from C2PSA which is last in the Backbone. I am using hook to get the output of C2PSA and passing it to down streaming task which is good.
But I am facing a problem in training the model as I am not using other layers and the model.train() is designed for the get trained on Bounding boxes data.
Is there a way we can achieve this like train the model from backbone to backward only or something like this.


The exact error I am getting is from ultralytics/engine/model.py Line 796 
Because I am not passing the dataset to yolo modal and still want to train it from gradients coming from the modal ahead of C2PSA layer.

**Comment by waqas-phc** [comment]
I also Tried with ( module 19 layer ) C3K2 layer from the Neck where C3K = True but getting the same error.

TO be more specific I am combining this layers output as Vision with the text modality coming from Bert and passing it to the Control model for robotic task.

( Thinking of using the cross attention as well between these 2 modalities.  What do you suggest ?)

Any suggestions and any sort of a way you can tell will be apricated.  I am really looking forward to your response.

**Comment by glenn-jocher** [comment]
It sounds like an interesting project! For your setup, consider using a custom training loop to handle gradients from your downstream task, bypassing YOLO's built-in training mechanism. Cross-attention could be a promising approach for integrating vision and text modalities.

**Comment by waqas-phc** [comment]
> It sounds like an interesting project! For your setup, consider using a custom training loop to handle gradients from your downstream task, bypassing YOLO's built-in training mechanism. Cross-attention could be a promising approach for integrating vision and text modalities.

Thanks I will try that

**Comment by glenn-jocher** [comment]
You're welcome! If you have any more questions as you proceed, feel free to ask.
